core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -303,19 +302,11 @@ public ShardRouting activeReplica(ShardRouting shard) {
     }
 
     /**
-     * Returns all shards that are not in the state UNASSIGNED with the same shard
-     * ID as the given shard.
-     */
-    public Iterable<ShardRouting> assignedShards(ShardRouting shard) {
-        return assignedShards(shard.shardId());
-    }
-
-    /**
      * Returns <code>true</code> iff all replicas are active for the given shard routing. Otherwise <code>false</code>
      */
-    public boolean allReplicasActive(ShardRouting shardRouting) {
-        final List<ShardRouting> shards = assignedShards(shardRouting.shardId());
-        if (shards.isEmpty() || shards.size() < this.routingTable.index(shardRouting.index().getName()).shard(shardRouting.id()).size()) {
+    public boolean allReplicasActive(ShardId shardId, int expectedAssignedShardsCount) {
```

Comment:
I rewrote the method, relying on index metadata to get the number of shards in the replica group instead.